firmware-management/swupdate-handlers/rootfs-handler.c

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <errno.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include "swupdate.h"
+#include "util.h"
+#include "handler.h"
+#include "arm-handler-common.h"
+
+int copy_image_and_sync(struct img_type *img, const char *const device_filepath)
+{
+    int ret_val = 0;
+    int fd = openfileoutput(device_filepath);
+    if (fd < 0)
+    {
+        ERROR("%s", "Failed to open target device file");
+        return -1;
+    }
+
+    if (copyimage(&fd, img, NULL) < 0)
+    {
+        ERROR("%s", "Failed to copy image to target device");
+        goto close;
+        ret_val = -1;
+    }
+
+    if (fsync(fd) == -1)
+    {
+        WARN("%s: %s", "Failed to sync filesystem", strerror(errno));
+    }
+
+close:
+    if (close(fd) == -1)
+    {
+        ERROR("%s", "Failed to close target device file descriptor");
+        ret_val = -1;
+    }
+
+    return ret_val;
+}
+
+int get_bootflag_file_path(char *const bootflags_file_path, const size_t size)
+{
+    static const char *const fname = "/rootfs2";
+    const size_t len = strlen(BOOTFLAGS_DIR) + strlen(fname) + 1;
+    if (len > size)
+    {
+        ERROR("%s", "Bootflag file path will not fit in the destination buffer");
+        return -1;
+    }
+
+    return snprintf(bootflags_file_path, len, "%s%s", BOOTFLAGS_DIR, fname);
+}
+
+int write_bootflag_file()
+{
+    if (mkdir(BOOTFLAGS_DIR, 0755) == -1)
+    {
+        if (errno != EEXIST)
+        {
+            ERROR("%s %s: %s", "Failed to create", BOOTFLAGS_DIR, strerror(errno));
+            return -1;
+        }
+    }
+
+    char bootflags_file_path[PATH_MAX];
+    if (get_bootflag_file_path(bootflags_file_path, PATH_MAX) < 1)
+        return -1;
+
+    FILE *const file = fopen(bootflags_file_path, "w");
+    if (!file)
+    {
+        ERROR("%s: %s", "Failed to open bootflags file", strerror(errno));
+        return -1;
+    }
+
+    fprintf(file, "%s", "");
```

Comment:
Do you need this? I would have thought the `fopen`/`fclose` would be enough. If it isn't then you can definitely touch a file just using Linux's `open`/`close`. 

I think we should `fsync` after creating the file... But will fsync force writes to the parent directory to be flushed, or just the file itself? Maybe `syncfs`?

Reply:
Yeah it's needed when using `fopen`, I'll add a `syncfs` as that will only synchronize the fs containing the file rather than flushing all buffered modifications to the fs like `sync` does.

Reply:
Thanks.